notebooks/official/ray_on_vertex_ai/train/model_training_formalization_with_ray_and_kfp.ipynb

```diff
@@ -0,0 +1,1765 @@
+        "@dsl.component(\n",
+        "    base_image=\"python:3.9\",\n",
+        "    packages_to_install=[\n",
+        "        \"kfp\",\n",
+        "        \"google-cloud-aiplatform\",\n",
+        "        \"google-cloud-pipeline-components\",\n",
+        "    ],\n",
+        ")\n",
+        "def GetDatasetUriOp(\n",
+        "    dataset: Input[VertexDataset], project: str, region: str, bucket_uri: str\n",
+        ") -> NamedTuple(\"outputs\", dataset_uri=str):\n",
+        "    \"\"\"Get the dataset resource name.\"\"\"\n",
+        "\n",
+        "    import logging\n",
+        "\n",
+        "    from google.cloud import aiplatform as vertex_ai\n",
+        "    from google.protobuf.json_format import MessageToDict\n",
+        "\n",
+        "    # Configure logging\n",
+        "    logging.basicConfig(\n",
+        "        level=logging.INFO, format=\"%(asctime)s - %(levelname)s - %(message)s\"\n",
+        "    )\n",
+        "\n",
+        "    # Initialize the Vertex AI SDK\n",
+        "    logging.info(\"Initializing Vertex AI SDK.\")\n",
+        "    vertex_ai.init(project=project, location=region, staging_bucket=bucket_uri)\n",
+        "\n",
+        "    # Get the dataset resource name\n",
+        "    logging.info(\n",
+        "        f\"Getting dataset resource name from Vertex AI dataset '{dataset.metadata['resourceName']}'.\"\n",
+        "    )\n",
+        "    vertex_ai_dataset = vertex_ai.TabularDataset(\n",
+        "        dataset_name=dataset.metadata[\"resourceName\"]\n",
+        "    )\n",
+        "    bq_uri = MessageToDict(vertex_ai_dataset.gca_resource._pb)[\"metadata\"][\n",
+        "        \"inputConfig\"\n",
+        "    ][\"bigquerySource\"][\"uri\"].replace(\"bq://\", \"\")\n",
+        "\n",
+        "    component_outputs = NamedTuple(\"outputs\", dataset_uri=str)\n",
+        "    return component_outputs(bq_uri)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "e-EUrudybvu9"
+      },
+      "source": [
+        "#### Build a `VertexRayJobOp` component to run training job using Ray Jobs API\n",
+        "\n",
+        "To orchestrate a Ray Job in a Vertex AI Pipeline, you need to create a KFP pipeline component which takes a Python script and submit the script to an existing Ray cluster on Vertex AI using the Ray Jobs API."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "1QvQM3wVHyNp"
+      },
+      "source": [
+        "##### Build the `VertexRayJobOp` component\n",
+        "\n",
+        "The `VertexRayJobOp` component submit the Ray Jobs API throught the public Ray dashboard address. **It is important to hightlight that  is accessible from outside the VPC, including the public internet. Use this component for experimentation only**. For production application, consider to set up connectivity from Vertex AI to your VPN and be sure to deploy the Ray on Vertex AI cluster in the same network."
```

Comment:
- incorrect: "submit the Ray Jobs API throught the public Ray dashboard address" (typo)
- should be: "submit the Ray Jobs API through the public Ray dashboard address"